bpf/ut/maps_test.go

```diff
@@ -90,6 +90,99 @@ func TestMapIterationDeleteAfter(t *testing.T) {
 	Expect(err2).NotTo(HaveOccurred(), "Failed to delete map entry")
 }
 
+// TestDeleteDuringIter tries to validate that BPF map iteration is working as we want even when keys are being added
+// to and deleted from the map during iteration.
+func TestDeleteDuringIter(t *testing.T) {
+	RegisterTestingT(t)
+	defer cleanUpMaps()
+	logLevel := logrus.GetLevel()
+	defer func() {
+		logrus.SetLevel(logLevel)
+	}()
+	logrus.SetLevel(logrus.InfoLevel)
+
+	testDelDuringIterN(10)
+	testDelDuringIterN(11)
+	testDelDuringIterN(100)
+	testDelDuringIterN(10000)
+	testDelDuringIterN(100000)
+}
+
+func testDelDuringIterN(numEntries int) {
+	cleanUpMaps()
+	for i := 0; i < numEntries; i++ {
+		err := insertNumberedKey(i)
+		Expect(err).NotTo(HaveOccurred())
+	}
+	// First pass, no deletions.
+	seenKeys := map[conntrack.Key]bool{}
+	err := ctMap.Iter(func(k, v []byte) bpf.IteratorAction {
+		key := conntrack.KeyFromBytes(k)
+		Expect(seenKeys[key]).To(BeFalse(), "Saw a duplicate key")
+		seenKeys[key] = true
+		return bpf.IterNone
+	})
+	Expect(err).NotTo(HaveOccurred())
+	Expect(seenKeys).To(HaveLen(numEntries), "Should have seen expected num entries on first iteration")
+
+	// Second pass, delete alternate keys.
+	seenKeys = map[conntrack.Key]bool{}
+	deleteNextKey := false
+	expectedKeys := map[conntrack.Key]bool{}
+	err = ctMap.Iter(func(k, v []byte) bpf.IteratorAction {
+		defer func() {
+			deleteNextKey = !deleteNextKey
+		}()
+		key := conntrack.KeyFromBytes(k)
+		Expect(seenKeys[key]).To(BeFalse(), "Saw a duplicate key")
+		seenKeys[key] = true
+		if deleteNextKey {
+			return bpf.IterDelete
+		} else {
+			expectedKeys[key] = true
+			return bpf.IterNone
+		}
+	})
+	Expect(err).NotTo(HaveOccurred())
+	Expect(seenKeys).To(HaveLen(numEntries), "Should have seen expected num entries on deletion iteration")
+
+	// Third pass, insert key on each iteration an delete on alternate iterations.
+	numLeftAfterSecondPass := len(expectedKeys)
+	seenKeys = map[conntrack.Key]bool{}
+	deleteNextKey = false
+	insertClock := 0
+	insertIdx := numEntries
```

Comment:
Does this mean we're only inserting at the end of the map?

Reply:
The maps are hashtables so I don't think they're ordered. The reason to use `insertIdx := numEntries` here is so that we get new entries rather than overwriting old ones.